File: python/src/a_mcp/ledger.py
"""Sealed records and the tamper-evident hash chain.

Mirrors the TypeScript ledger. A sealed record is an emitted event plus host-assigned fields
(sequence, prev_hash, record_hash). In a real deployment these tiers map to DynamoDB (Tier1
durable accept) and a sealing step writing to S3 Object Lock (Tier2 WORM + hash-chain); here
both are simulated locally and deterministically, because the standardization-relevant
artifact is the hash chain, not the infrastructure.
"""

from dataclasses import dataclass

from a_mcp.canonical import canonicalize, sha256_hex

# The genesis link the first record chains from.
GENESIS_HASH = '0' * 64


def compute_record_hash(event: dict, seq: int, host_ts: str, prev_hash: str) -> str:
    """Compute the record hash over the committed bytes.

    Kept explicit so an independent verifier (and the Python/TypeScript ports) recompute it
    identically: ``sha256( canonical(event) | seq | host_ts | prev_hash )``.

    Args:
        event: The audit event with absent optionals already omitted.
        seq: The partition-monotonic sequence.
        host_ts: The authoritative host timestamp.
        prev_hash: The previous record hash in the chain.

    Returns:
        The hex-encoded SHA-256 record hash.
    """
    preimage = f'{canonicalize(event)}|{seq}|{host_ts}|{prev_hash}'
    return sha256_hex(preimage)
    # end def


@dataclass
class SealedRecord:
    """A tool-emitted event plus the host-assigned ledger fields."""

    event: dict
    seq: int
    host_ts: str
    prev_hash: str
    record_hash: str
    # end class


class Ledger:
    """Append-only, per-partition tamper-evident ledger."""

    def __init__(self, partition: str) -> None:
        """Initialize an empty ledger for the given partition."""
        self.partition = partition
        self._records: list[SealedRecord] = []
        # end def

    def append(self, event: dict, host_ts: str) -> SealedRecord:
        """Append an event, assigning the next sequence and linking the hash chain."""
        seq = len(self._records)
        prev_hash = self._records[-1].record_hash if self._records else GENESIS_HASH
        record_hash = compute_record_hash(event, seq, host_ts, prev_hash)
        sealed = SealedRecord(event=event, seq=seq, host_ts=host_ts, prev_hash=prev_hash, record_hash=record_hash)
        self._records.append(sealed)
        return sealed
        # end def

    def records(self) -> list[SealedRecord]:
        """Return all sealed records in append order."""
        return list(self._records)
        # end def

    def digest(self) -> str:
        """Return the anchored digest (the tail record hash, or genesis if empty)."""
        return self._records[-1].record_hash if self._records else GENESIS_HASH
        # end def

    # end class

File: typescript/src/ledger/ledger.ts
import type { AuditEvent } from '../schema/event.js';
import { canonicalize, sha256Hex } from './canonical.js';

// Sealed record = tool-emitted event + host-assigned ledger fields. In a real deployment
// these tiers map to DynamoDB (Tier1 durable accept) + Sealing Lambda → S3 Object Lock
// (Tier2 WORM + hash-chain), design audit-integrity §2. The PoC simulates both locally and
// deterministically — the standardization-relevant artifact is the hash chain, not AWS.

export const GENESIS_HASH = '0'.repeat(64);

export interface SealedRecord {
  event: AuditEvent;
  seq: number; // partition-monotonic sequence assigned by the sealer (a gap means a record was lost)
  host_ts: string; // authoritative host time
  prev_hash: string; // previous record_hash in the chain
  record_hash: string; // sha256( canonical(event) ‖ seq ‖ host_ts ‖ prev_hash )
}

// The bytes the chain commits to. Kept explicit so the verifier recomputes identically.
export function computeRecordHash(
  event: AuditEvent,
  seq: number,
  hostTs: string,
  prevHash: string,
): string {
  const preimage = `${canonicalize(event)}|${seq}|${hostTs}|${prevHash}`;
  return sha256Hex(preimage);
}

// Append-only, per-partition ledger. A partition is a tenant#day-like scope; the PoC keeps
// one in memory. Append assigns the next seq and links the hash chain.
export class Ledger {
  private records: SealedRecord[] = [];

  constructor(public readonly partition: string) {}

  append(event: AuditEvent, hostTs: string): SealedRecord {
    const seq = this.records.length;
    const prev = this.records[seq - 1];
    const prevHash = prev ? prev.record_hash : GENESIS_HASH;
    const record_hash = computeRecordHash(event, seq, hostTs, prevHash);
    const sealed: SealedRecord = { event, seq, host_ts: hostTs, prev_hash: prevHash, record_hash };
    this.records.push(sealed);
    return sealed;
  }

  all(): readonly SealedRecord[] {
    return this.records;
  }

  // Digest anchored out-of-band (Tier3). Rewriting history without invalidating this is
  // infeasible because the tail commits to the whole chain.
  digest(): string {
    const tail = this.records[this.records.length - 1];
    return tail ? tail.record_hash : GENESIS_HASH;
  }

  // Test/demo affordance: expose the mutable array so a tamper can be injected and caught.
  unsafeMutableRecords(): SealedRecord[] {
    return this.records;
  }
}
